ios/Classes/ComWilliamrijksenOnesignalModule.m

```diff
@@ -76,35 +87,141 @@ - (void) actionHandler:(OSNotificationOpenedResult *)result {
-          }
+                                    NSLog(@"[DEBUG] com.williamrijksen.onesignal notification");
+                                    [self receivedHandler:notification];
+                                }
             handleNotificationAction:^(OSNotificationOpenedResult *result) {
-                [self actionHandler:result];
-            }
-                            settings:@{
-                 kOSSettingsKeyInFocusDisplayOption: @(OSNotificationDisplayTypeNone),
-                 kOSSettingsKeyAutoPrompt: @YES}
-     ];
-	//TODO these settings should be configurable from the Titanium App on module initialization
+                                    NSLog(@"[DEBUG] com.williamrijksen.onesignal User opened notification");
+                                    [self actionHandler:result];
+                                }
+                            settings:@{kOSSettingsKeyAutoPrompt: @false}
+    ];
+}
+
++ (void)load
+{
+    NSLog(@"[DEBUG] com.williamrijksen.onesignal load");
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(onAppCreate:)
+                                                 name:@"UIApplicationDidFinishLaunchingNotification" object:nil];
+}
+
+- (void)handleQueuedBootEvent:(NSString*)eventName
+{
+    if (_queuedBootEvents != nil && [_queuedBootEvents objectForKey:eventName] != nil) {
+        NSLog(@"[DEBUG] com.williamrijksen.onesignal Fire event from startup %@", eventName);
+        [[NSNotificationCenter defaultCenter] postNotificationName:eventName object:self userInfo:[_queuedBootEvents objectForKey:eventName]];
+        [_queuedBootEvents removeObjectForKey:eventName];
+    }
+}
+
+-(void)_listenerAdded:(NSString*)type count:(int)count
+{
+    NSLog(@"[DEBUG] com.williamrijksen.onesignal add listener %@", type);
+    if (count == 1 && [type isEqual:TiNotificationOpened]) {
+        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(notificationOpened:) name:TiNotificationOpened object:nil];
+        [self handleQueuedBootEvent:type];
+    }
+    if (count == 1 && [type isEqual:TiNotificationReceived]) {
+        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(notificationReceived:) name:TiNotificationReceived object:nil];
+        [self handleQueuedBootEvent:type];
+    }
+}
+
+-(void)_listenerRemoved:(NSString*)type count:(int)count
+{
+    NSLog(@"[DEBUG] com.williamrijksen.onesignal remove listener");
+    if (count == 0 && [type isEqual:TiNotificationOpened]) {
+        [[NSNotificationCenter defaultCenter] removeObserver:self name:TiNotificationOpened object:nil];
+    }
+    if (count == 0 && [type isEqual:TiNotificationReceived]) {
+        [[NSNotificationCenter defaultCenter] removeObserver:self name:TiNotificationReceived object:nil];
+    }
+}
+
+-(void)notificationOpened:(NSNotification*)info
+{
+    if (_instance != nil) {
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
```

Comment:
@hansemannn is this a method or something else? I can't find anything in the titanium_mobile repository with the name `TiDispatchOnMainThread`.